[16.x] Improve sysroot detection on MacOS

vepadulano opened the pull request:
Forward-port of a patch present in 5.x and 9.x clang release cycles, see:

https://github.com/conda-forge/clangdev-feedstock/commit/027fdaaa7b27af70f74ea9e54397facaaebfb5aa

<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

h-vetinari commented:
This will need proper rebasing on a local llvm-project 16.x branch, with all the existing patches applied first, e.g. through
```
git clone https://github.com/llvm/llvm-project
cd llvm-project
git checkout tags/llvmorg-16.0.6
find ../path/to/clangdev-feedstock/recipe/patches/ | xargs git am      # feedstock checked out on branch 16.x
# apply, fix & commit the patch being added here, keep metadata (author, date)
git format-patch tags/llvmorg-16.0.6 --no-signature
```

Also, do you want to merge #294 before this one, after this one, or together?

vepadulano commented:
Thanks, yes indeed it needs to be properly rebased. IMHO probably we want the other PR merged first, let me double check once more that it works for the ROOT patch as well and then we can proceed with merging.

h-vetinari reviewed (approved):
From my POV this looks alright, but since this is changing also our non-root production compilers to look into `CONDA_BUILD_SYSROOT` before `SDKROOT`, it would be good for @isuruf to comment as well.

isuruf commented on recipe/patches/0010-Improve-logic-for-finding-the-macos-sysroot-for-cond.patch:
```diff
@@ -0,0 +1,107 @@
+From 06e16a989f21d8c9ec0382a091f0dd32cb83b1da Mon Sep 17 00:00:00 2001
+From: Chris Burr <christopher.burr@cern.ch>
+Date: Sat, 17 Apr 2021 09:26:17 +0200
+Subject: [PATCH 10/10] Improve logic for finding the macos sysroot for conda
```
Can you merge this patch with `recipe/patches/0002-clang-add-conda-specific-env-var-CONDA_BUILD_SYSROOT.patch` and also check SDKROOT before CONDA_BUILD_SYSROOT ?

h-vetinari commented on recipe/patches/0002-clang-add-conda-specific-env-var-CONDA_BUILD_SYSROOT.patch:
```diff
@@ -1,12 +1,18 @@
+
+Co-authored-by: Chris Burr <christopher.burr@cern.ch>
+Co-authored-by: Vincenzo Eduardo Padulano <v.e.padulano@gmail.com>
+Co-authored-by: Fernando Isuru <isuruf@gmail.com>
```
It's the other way around.

vepadulano replied:
Right, my bad, apologies for that!

h-vetinari replied:
Please fix it (also in the commit message of https://github.com/conda-forge/clangdev-feedstock/pull/298/commits/a3dcec494e0d298e1d3baee7f01aa7bb495e828c)

h-vetinari commented on recipe/patches/0001-Fix-sysroot-detection-for-linux.patch:
```diff
@@ -1,4 +1,4 @@
-From e46d9dd8f6701270b5243348ffb4b8c3430c4229 Mon Sep 17 00:00:00 2001
+From eaf10f5bfe63fa3279f193ea9f95671970e8282b Mon Sep 17 00:00:00 2001
 From: Isuru Fernando <isuruf@gmail.com>
```
See here, for example

h-vetinari commented:
Also, there seems to be an issue with the implementation on windows:
```
[461/1896] Building CXX object lib\Driver\CMakeFiles\obj.clangDriver.dir\ToolChains\Darwin.cpp.obj
FAILED: lib/Driver/CMakeFiles/obj.clangDriver.dir/ToolChains/Darwin.cpp.obj 
C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\HostX64\x64\cl.exe  /nologo /TP -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I%SRC_DIR%\clang\build\lib\Driver -I%SRC_DIR%\clang\lib\Driver -I%SRC_DIR%\clang\include -I%SRC_DIR%\clang\build\include -I%PREFIX%\Library\include -MD  /DWIN32 /D_WINDOWS   /Zc:inline /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd4324 -w14062 -we4238 /Gw /MD /O2 /Ob2 /DNDEBUG -std:c++17  /EHs-c- /GR /showIncludes /Folib\Driver\CMakeFiles\obj.clangDriver.dir\ToolChains\Darwin.cpp.obj /Fdlib\Driver\CMakeFiles\obj.clangDriver.dir\ /FS -c %SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2134): error C2065: 'pclose': undeclared identifier
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2134): error C3861: 'popen': identifier not found
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2134): error C2065: 'pclose': undeclared identifier
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2134): error C2641: cannot deduce template arguments for 'std::unique_ptr'
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2135): error C2451: a conditional expression of type 'std::unique_ptr' is not valid
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2135): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2136): error C2662: '_Get_deleter_pointer_type<_Ty,remove_reference<_Ty1>::type,void>::type std::unique_ptr<_Ty,_Dx>::get(void) noexcept const': cannot convert 'this' pointer from 'std::unique_ptr' to 'const std::unique_ptr<_Ty,_Dx> &'
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2136): note: Reason: cannot convert from 'std::unique_ptr' to 'const std::unique_ptr<_Ty,_Dx>'
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2136): note: Conversion requires a second user-defined-conversion operator or constructor
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.29.30133\include\memory(3218): note: see declaration of 'std::unique_ptr<_Ty,_Dx>::get'
%SRC_DIR%\clang\lib\Driver\ToolChains\Darwin.cpp(2136): error C2660: 'fgets': function does not take 2 arguments
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\stdio.h(198): note: see declaration of 'fgets'
```
I think that `popen` / `pclose` need a leading underscore on windows, c.f. [here](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/pclose?view=msvc-170)

vepadulano commented:
> I think that popen / pclose need a leading underscore on windows, c.f. [here](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/pclose?view=msvc-170)

The MSVC doc state `This API cannot be used in applications that execute in the Windows Runtime.` and the extra patch is really only relevant for MacOS. I am unsure on how to proceed: either we keep the two patches merged and add some preprocessor macro to avoid using the unsupported API on Windows or we split again tha patches and have the MacOS one protected by a selector. @isuruf what do you think?

h-vetinari commented:
You could also hide the extra code behind `#ifndef _WIN32` or perhaps `#ifndef _MSC_VER`

isuruf commented:
You can copy over https://github.com/llvm/llvm-project/blob/ce863dd5a77723fc21e1e43f5d3dd503e9a73d14/clang-tools-extra/clangd/CompileCommands.cpp#L120-L131 and https://github.com/llvm/llvm-project/blob/ce863dd5a77723fc21e1e43f5d3dd503e9a73d14/clang-tools-extra/clangd/CompileCommands.cpp#L40-L76

h-vetinari reviewed (approved):
Reconfirm

isuruf commented on recipe/patches/0002-clang-add-conda-specific-env-var-CONDA_BUILD_SYSROOT.patch:
```diff
@@ -24,6 +30,105 @@ index a268f2fa8fc5..baf84d4d1adb 100644
++      }
++    }
++
++    // Check if the command line tools are installed in the default directory
++    prefixToTry = "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk";
++    if (foundSDKPath == "/" && getVFS().exists(prefixToTry+"/usr/include/sys/types.h")) {
++      foundSDKPath = std::string(prefixToTry);
++    }
++
++    // Check if xcode is installed in the default directory
++    prefixToTry = "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk";
++    if (foundSDKPath == "/" && getVFS().exists(prefixToTry+"/usr/include/sys/types.h")) {
++      foundSDKPath = std::string(prefixToTry);
++    }
```
Move these to the end as they hard code `MacOSX.sdk` which prevents other darwin based OSes.

vepadulano commented on recipe/patches/0002-clang-add-conda-specific-env-var-CONDA_BUILD_SYSROOT.patch:
```diff
@@ -24,6 +30,105 @@ index a268f2fa8fc5..baf84d4d1adb 100644
++      // Run xcode-select, based on https://stackoverflow.com/a/478960
++      std::array<char, 128> buffer;
++      std::string result;
++      std::unique_ptr<FILE, decltype(&pclose)> pipe(popen("xcode-select -p", "r"), pclose);
```
Just to check, do you mean to also include `xcrun` as a fallback in case `xcode-select` doesn't work or to use it instead of it? I'm asking because on my Mac the two executables report different paths and both seem valid:
```
$:  xcrun --show-sdk-path
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
$: ls /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/types.h
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/types.h
$:  xcode-select -p
/Applications/Xcode.app/Contents/Developer
$: ls /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/types.h
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/types.h
```

vepadulano commented:
@isuruf I moved the MacOS-specific checks at the end and I implemented the lookup using CLI, `xcrun` first and `xcode-select` second. Let me know if it looks good or I should remove the `xcode-select` part completely

isuruf commented on recipe/patches/0002-clang-add-conda-specific-env-var-CONDA_BUILD_SYSROOT.patch:
```diff
@@ -24,6 +30,141 @@ index a268f2fa8fc5..baf84d4d1adb 100644
++      if (foundSDKPath == "/" && XcrunPath &&
++          getVFS().exists(XcrunPath.value() + "/usr/include/sys/types.h")) {
++        foundSDKPath = XcrunPath.value();
++      } else {
++        auto XcodeSelectPath = pathFromExecutable({"xcode-select", "-p"});
++        if (XcodeSelectPath) {
++          const auto &result = XcodeSelectPath.value();
++          // Check if the command line tools are installed
++          prefixToTry = result + "/SDKs/MacOSX.sdk";
++          if (foundSDKPath == "/" &&
++              getVFS().exists(prefixToTry + "/usr/include/sys/types.h")) {
++            foundSDKPath = prefixToTry;
++          }
++          // Check if xcode is installed
++          prefixToTry =
++              result + "/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk";
++          if (foundSDKPath == "/" &&
++              getVFS().exists(prefixToTry + "/usr/include/sys/types.h")) {
++            foundSDKPath = prefixToTry;
++          }
++        }
++      }
++    }
```
Let's remove this entire section.

vepadulano commented:
HI @isuruf, sorry for bothering, do you think this PR can be merged? Otherwise let me know how I should improve it, thanks!

isuruf commented on recipe/patches/0002-clang-add-conda-specific-env-var-CONDA_BUILD_SYSROOT.patch:
```diff
@@ -24,6 +30,122 @@ index a268f2fa8fc5..baf84d4d1adb 100644
++      }
++    }
++
++    // Check if the command line tools are installed in the default directory
++    prefixToTry = "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk";
++    if (foundSDKPath == "/" &&
++        getVFS().exists(prefixToTry + "/usr/include/sys/types.h")) {
++      foundSDKPath = prefixToTry;
++    }
++
++    // Check if xcode is installed in the default directory
++    prefixToTry = "/Applications/Xcode.app/Contents/Developer/Platforms/"
++                  "MacOSX.platform/Developer/SDKs/MacOSX.sdk";
++    if (foundSDKPath == "/" &&
++        getVFS().exists(prefixToTry + "/usr/include/sys/types.h")) {
++      foundSDKPath = prefixToTry;
++    }
```
Let's remove these

vepadulano replied:
Done!

isuruf commented:
Thanks. We should add this to newer versions too.

h-vetinari commented:
> We should add this to newer versions too.

I've been waiting for this merge to fold this into #297
